File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.py
from collections import Counter
def solution(weights):
    answer = 0
    # 1:1
    counter = Counter(weights)
    for k,v in counter.items():
        # 같은게 2개 이상있을 때 
        if v>=2:
            answer+= v*(v-1)//2
    weights = set(weights) 
    # print(weights)
    # print(counter)
    
    # 2:3 2:4 3:4 
    for w in weights:
        if w*2/3 in weights:
            answer+= counter[w*2/3] * counter[w]
        if w*2/4 in weights:
            answer+= counter[w*2/4] * counter[w]
        if w*3/4 in weights:
            answer+= counter[w*3/4] * counter[w]
    return answer

File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.js
function solution(weights) {
    var answer = 0;
    const ratios = [
        [1,1],
        [3,2],
        [4,3],
        [2,1],
        [2,3],
        [3,4],
        [1,2],
    ]
    
    const countMap = new Map()
    for(let w of weights){
        for(const [a, b] of ratios){
            const target = (w * a) / b;
            if(Number.isInteger(target) && countMap.has(target)){
                answer += countMap.get(target)
            }
        }
        
        countMap.set(w, (countMap.get(w) ??  0) + 1)
    }
    
    return answer;
}

  // const countMap = new Map()
    // 몸무게별 인원수 저장
//     for(let weight of weights){
//         countMap.set(weight, (countMap.get(weight) || 0) + 1)
//     }
    
    // 100, 100, 100, 100, 100 -> 10
    // 100, 100, 100, 100 -> 6
    // 100, 100, 100 -> 3
    // n(n-1)/2
    // 같은 사람 몸무게 먼저 쌍찾기
//     for(let [weight, cnt] of wMap.entries()){
//         if(cnt >= 2){
//             answer += cnt * (cnt-1) / 2 
//         }
